android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/RibEvents.kt

```diff
@@ -20,24 +20,96 @@ import kotlinx.coroutines.channels.BufferOverflow
+public object RibEvents {
 
-  private val _events = MutableSharedFlow<RibEvent>(0, 1, BufferOverflow.DROP_OLDEST)
-  public val events: Observable<RibEvent> = _events.asObservable()
+  private val mutableRouterEvents =
+    MutableSharedFlow<RibRouterEvent>(0, 1, BufferOverflow.DROP_OLDEST)
+  private val mutableRibDurationEvents =
+    MutableSharedFlow<RibActionInfo>(0, 1, BufferOverflow.DROP_OLDEST)
+
+  @JvmStatic
+  public val routerEvents: Observable<RibRouterEvent> = mutableRouterEvents.asObservable()
+
+  @JvmStatic
+  public val ribActionEvents: Observable<RibActionInfo> = mutableRibDurationEvents.asObservable()
 
   /**
    * @param eventType [RibEventType]
    * @param child [Router]
    * @param parent [Router] and null for the root ribs that are directly attached to
    *   RibActivity/Fragment
    */
-  public fun emitEvent(eventType: RibEventType, child: Router<*>, parent: Router<*>?) {
-    _events.tryEmit(RibEvent(eventType, child, parent))
+  public fun emitRouterEvent(eventType: RibEventType, child: Router<*>, parent: Router<*>?) {
+    mutableRouterEvents.tryEmit(RibRouterEvent(eventType, child, parent))
+  }
+
+  /**
+   * Calls related RIB action (e.g. didBecomeActive) and emits emission of ATTACHED/DETACHED events
+   * for each RIB component.
+   *
+   * @param ribAction The related RIB action type. e.g. didBecomeActive, willLoad, etc
+   * @param ribCallerClassType Related RIB component class type
+   * @param ribComponentType The RIB component type (e.g. Interactor, Router, Presenter, Worker)
+   * @param ribEventType RIB event type (e.g. ATTACH/DETACH)
+   */
+  internal inline fun <T : Any> triggerRibActionAndEmitEvents(
+    ribCallerClassType: T,
+    ribComponentType: RibComponentType,
+    ribEventType: RibEventType,
+    ribAction: () -> Unit,
+  ) {
+    val ribClassName = ribCallerClassType.asQualifiedName()
+    ribClassName?.emitRibEventAction(ribComponentType, ribEventType, RibActionState.STARTED)
+    ribAction()
+    ribClassName?.emitRibEventAction(ribComponentType, ribEventType, RibActionState.COMPLETED)
   }
 
-  public companion object {
-    private val instance: RibEvents = RibEvents()
+  private fun Any.asQualifiedName(): String? = javaClass.kotlin.qualifiedName
 
-    @JvmStatic public fun getInstance(): RibEvents = instance
+  /**
+   * Emits emission of ATTACHED/DETACHED events for each RIB component.
+   *
+   * @param ribComponentType The RIB component type (e.g. Interactor, Router, Presenter)
+   * @param ribEventType RIB event type (e.g. ATTACH/DETACH)
+   * @param ribActionState: RibActionType,
+   */
+  private fun String.emitRibEventAction(
+    ribComponentType: RibComponentType,
+    ribEventType: RibEventType,
+    ribActionState: RibActionState,
+  ) {
+    val ribActionInfo =
+      RibActionInfo(
+        this,
+        ribComponentType,
+        ribEventType,
+        ribActionState,
+        Thread.currentThread().name,
+      )
+    mutableRibDurationEvents.tryEmit(ribActionInfo)
```

Comment:
Maybe we want to take an explicit flag that enables event callbacks (or check for subcribers) to avoid creating objects if no one is listening.

Reply:
Added a factory `RibEvents.startCapturingRibActionInfo()` to be called explicitly before any ribActionEvents collection